### PR TITLE
Allow specifying custom HTTP headers in WebSocket connect_to_url

### PR DESCRIPTION
**Pull request by anoadragon453**
This PR allows the user to specify a dictionary of custom HTTP headers that will be used during the WebSocket HTTP connection upgrade handshake.

An example of usage:

```
var headers = {"Cookie": "login_token=1JtHVmhWqmvK6KLbVcQ8RT2oEhW9TLAmRv"}
var err = _ws_client.connect_to_url(ws_url, PoolStringArray(), false, headers)
```

Due to limitations in JavaScript, this feature is only available to the native C++ client.

### CODE REVIEW

**Comment by anoadragon453** [comment]
Not sure what the release cadence is and/or whether you're in feature freeze at the moment, however @Faless mentioned wanting to rework the WebSocket module for 3.2, which would make this PR obsolete.

Is it possible to thus have this for the 3.1 merge window instead?

**Comment by akien-mga** [comment]
We are in feature freeze, so a priori no. But it's up to @Faless as maintainer, if he thinks this can go in 3.1, he can merge it.

**Review by Faless — CHANGES_REQUESTED** [review]
The binding part also needs to be updated:
https://github.com/godotengine/godot/blob/master/modules/websocket/websocket_client.cpp#L134
```
ClassDB::bind_method(D_METHOD("connect_to_url", "url", "protocols", "gd_mp_api", "custom_headers"), &WebSocketClient::connect_to_url, DEFVAL(PoolVector<String>()), DEFVAL(false), DEFVAL(Dictionary()));
```

---

Related to #27129.

Sending custom headers will never work in HTML5 exports as it's not part of the [living standard](https://html.spec.whatwg.org/multipage/web-sockets.html).

The only exception is cookies (which browsers send by default if set for the given domain).
Looking around people seems to either send authinfo via the path/query part of the URL or via cookies (which are though limited by the domain name).

I **would advise users use the more compatible query string** instead (maybe adding its retrieval to WebSocketServer).
But I would like to hear some feedback (downsides of using query string?), I don't have a strong opinion on this.

**Comment by akien-mga** [comment]
Could you do the requested changes?

**Comment by anoadragon453** [comment]
Oh sorry, I was a little confused on whether this PR would get accepted at all as I thought we were switching to a different library for the next Godot version - but is that not the case anymore?

**Comment by Faless** [comment]
> I thought we were switching to a different library for the next Godot version - but is that not the case anymore?

Hi, I'm sorry. I thought the switch to the new library would be done by now, but the work on WebRTC took a bit more time then expected.

We still plan to do the switch, but it might take too long to be safely included in `3.2`.

If you really care about this feature in `3.2` and wish to go ahead and rebase/fix this PR, it will be accepted (after additional code review if necessary).

If the library switch gets merged before, it will in any case include this feature.

I'm sorry about this situation, things don't always go as planned.

**Comment by anoadragon453** [comment]
@Faless No worries, and thanks for the context. I'll give it a look today :)

**Comment by anoadragon453** [comment]
Well, err, needless to say I never gave this a look :/

But, looks like things have moved on to the wslay library now in https://github.com/godotengine/godot/pull/30263, so this PR is likely no longer necessary?

**Pull request closed without merging**
